pkg/operator/k8sutil/resources.go

```diff
@@ -24,8 +24,55 @@ import (
 	"github.com/ghodss/yaml"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
+// OwnerInfo is to set owner references. Only one of owner and ref must be valid.
+type OwnerInfo struct {
+	owner    metav1.Object
+	scheme   *runtime.Scheme
+	ownerRef *metav1.OwnerReference
+}
+
+// NewOwnerInfo create a new ownerInfo to set ownerReference by controllerutil
+func NewOwnerInfo(owner metav1.Object, scheme *runtime.Scheme) *OwnerInfo {
```

Comment:
Tell me again why the scheme isn't always available with the owner reference? Is it when we're running inside the osd prepare job for example? Seems like everywhere else should have the scheme.

Reply:
There are two cases where scheme isn't available. One is osd preparae job as you said.

https://github.com/rook/rook/blob/master/cmd/rook/ceph/osd.go#L231

The other is in updating CSI driver.

https://github.com/rook/rook/blob/master/pkg/operator/ceph/operator.go#L243

Here the owner is the operator's deployment and the child is the CSI driver's configmap.

My colleague told me that we can create scheme by ourselves with `NewScheme()` and  the subsequent `AddToScheme()`s. However, it's a bit complecated and would take a bit long time. So, how about adopting the current logic for now and eliminate `k8sutil.SetOwnerRef()` after that?

Reply:
In two steps is fine. The key now is to have thorough testing to prevent regressions. With such a big change to the owner references, let's also test by ensuring the K8s garbage collector triggers rebuilding its index. This will help us verify resources don't later disappear 24 hours after everything is fine. However, I don't recall how to force trigger the GC rebuild.

Reply:
> let's also test by ensuring the K8s garbage collector triggers rebuilding its index.

Let me clarify what you mean. Isn't it enough to already merged `kubectl-check-ownerreferences`. If there are invalid ownerReferences, this program detect it. Na resource deletion only happenss due to invalid references.

>  I don't recall how to force trigger the GC rebuild.

Me neither. I guess how to force trigger the GC rebuild depends on implementation, so K8s version. So it's dufficult to do it.

Reply:
> > let's also test by ensuring the K8s garbage collector triggers rebuilding its index.
> 
> Let me clarify what you mean. Isn't it enough to already merged `kubectl-check-ownerreferences`. If there are invalid ownerReferences, this program detect it. Na resource deletion only happenss due to invalid references.

Oh right, that CI action will be perfect to detect the issues. I haven't seen that action fail yet in any PRs. Have you seen it fail? If we could just double check that the CI action fails when there is a real failure, I think we're covered, thanks.

Reply:
> Have you seen it fail?

No. After making this PR ready to be reviewed, I'll check it again.